### PR TITLE
restore: fix several bugs related to column permutations

### LINKED ISSUE #438: restore csv files failed with tidb backend
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.
Use lightning to restore some csv files. 

Table schema:
```
CREATE TABLE `test` (
	`id` int(11) NOT NULL,
	`contract_no` varchar(64) DEFAULT NULL,
	`fund_seq_no` varchar(64) DEFAULT NULL,
	`term_no` int(11) DEFAULT NULL,
	`contract_type` varchar(8) DEFAULT NULL,
	`internal_transfer_tag` varchar(8) DEFAULT NULL,
	`prin_amt` int(11) DEFAULT NULL,
	`start_date` varchar(8) DEFAULT NULL,
	`end_date` varchar(8) DEFAULT NULL,
	`batch_date` varchar(8) DEFAULT NULL,
	`crt_time` timestamp DEFAULT CURRENT_TIMESTAMP,
	`region_code` varchar(8) DEFAULT NULL,
	`credit_code` varchar(64) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE COLUMNS(batch_date) (
	PARTITION `P20200224` VALUES LESS THAN ("2020-02-05 00:00:00"),
	PARTITION `P20200324` VALUES LESS THAN ("2020-03-05 00:00:00"),
	PARTITION `P20200424` VALUES LESS THAN ("2020-04-05 00:00:00"),
	PARTITION `P20200524` VALUES LESS THAN ("2020-05-05 00:00:00"),
	PARTITION `P_MAXVALUE` VALUES LESS THAN MAXVALUE
);
```

Some sample rows in the csv files:
```
contract_no,fund_seq_no,term_no,contract_type,internal_transfer_tag,prin_amt,start_date,end_date,region_code,credit_code
2020061030595,20200629,1,01,N,356,20200210,20200720,000000,
202023462,00628,1,01,N,3561,20200310,20200720,000001,
64030597,0627,1,01,N,3562,20200410,20200720,000002,33
64030598,626,12,02,Y,26368,20200510,20200620,000003,
030599,9520200625,1,01,N,3960,20200610,20200720,000005,999 
```

2. What did you expect to see?
Lightning should import data successfully.

3. What did you see instead?
Failed with following error:
image

4. Versions of the cluster
v4.0.7 

5. Operation logs
   
   
6. Configuration of the cluster and the task
lightning config:
image

7. Screenshot/exported-PDF of Grafana dashboard or metrics' graph in Prometheus for TiDB-Lightning if possible

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
Fix follow bugs:
- load wrong number of column permutation for file checkpoint
- incorrect column names fetch from column permutation, this will cause tidb encoded generates row table columns if the csv head order is different from table schema
- tidb encoder use wrong field info to encode field

close #438 

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Side effects

Related changes

### CODE REVIEW

**Comment by kennytm** [comment]
the new test failed

```
[2020-10-31T08:52:30.160Z] ERROR 1064 (42000) at line 1: You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 87 near "'select fund_seq_no, region_code, credit_code from test.test_perm WHERE contract_no = '2020061000019011020164030597';" 
```

File: tests/column_permutation/data/test.test_perm-schema.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	`crt_time` timestamp DEFAULT CURRENT_TIMESTAMP,
+	`region_code` varchar(8) DEFAULT NULL,
+	`credit_code` varchar(64) DEFAULT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
+PARTITION BY RANGE COLUMNS(batch_date) (
+	PARTITION `P20200224` VALUES LESS THAN ("2020-02-05 00:00:00"),
+	PARTITION `P20200324` VALUES LESS THAN ("2020-03-05 00:00:00"),
+	PARTITION `P20200424` VALUES LESS THAN ("2020-04-05 00:00:00"),
+	PARTITION `P20200524` VALUES LESS THAN ("2020-05-05 00:00:00"),
+	PARTITION `P_MAXVALUE` VALUES LESS THAN ("MAXVALUE")
```
Comment (kennytm) [inline tests/column_permutation/data/test.test_perm-schema.sql:21]:
```suggestion
	PARTITION `P_MAXVALUE` VALUES LESS THAN MAXVALUE
```

File: tests/column_permutation/data/test-schema-create.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+CREATE DATABASE `test` IF NOT EXISTS;
```
Comment (kennytm) [inline tests/column_permutation/data/test-schema-create.sql:1]: please use a less generic database name

**Review by kennytm — COMMENTED** [review]
rest LGTM

**Comment by overvenus** [comment]
LGTM

**Review by overvenus — APPROVED** [review]

**Pull request merged**